hilti/toolchain/src/compiler/unit.cc

```diff
@@ -399,322 +282,92 @@ Result<CxxCode> Unit::cxxCode() const {
-                    for ( auto i : v.walk(md->root()) ) {
-                        if ( auto x = v.dispatch(i); ! (x && *x) )
-                            continue;
-
-                        if ( auto cached = context->lookupModule(n.id()) ) {
-                            cached->requires_compilation = true;
-                            context->updateModule(*cached);
-                            break;
-                        }
-                    }
-                }
-            }
+    auto v = Visitor();
+    for ( auto i : v.walk(*_module) ) {
+        if ( auto rc = v.dispatch(i) ) {
+            if ( rc && *rc )
+                return true;
         }
-    };
-
-    // Run the visitors.
-    auto v = VisitorImports(context(), _modules);
-    for ( auto i : v.walk(module) )
-        v.dispatch(i);
-}
-
-bool Unit::_validateASTs(std::vector<std::pair<ID, NodeRef>>& modules,
-                         const std::function<bool(const ID&, NodeRef&)>& run_hooks_callback) {
-    if ( options().skip_validation )
-        return true;
-
-    auto valid = true;
-
-    for ( auto& [id, module] : modules ) {
-        if ( ! _validateAST(id, NodeRef(module), run_hooks_callback) )
-            valid = false;
     }
 
-    return valid;
+    return false;
 }
 
-/**
- * Recursive helper function to traverse the AST and collect relevant errors.
- * We pick errors on child nodes first, and then hide any further ones
- * located in parents along the way. We take only the first error in each
- * priority class (normal & low). If a node doesn't have a location, we
- * substitute the closest parent location.
- *
- * @param n root node for validation
- * @param closest_location location closest to *n* on the path leading to it
- * @param errors errors recorded for reporting so far; function will extend this
- * @return two booleans with the 1st indicating if we have already found a
- * normal priroity error on the path, and the 2nd if we have already found a
- * low priority error on the path.
- */
-static std::pair<bool, bool> _recursiveValidateAST(const Node& n, Location closest_location,
-                                                   std::vector<node::Error>* errors) {
-    auto have_normal = false;
-    auto have_low = false;
-
+static node::ErrorPriority _recursiveValidateAST(const Node& n, Location closest_location, node::ErrorPriority prio,
+                                                 int level, std::vector<node::Error>* errors) {
     if ( n.location() )
         closest_location = n.location();
 
-    for ( const auto& c : n.childs() ) {
-        auto [normal, low] = _recursiveValidateAST(c, closest_location, errors);
-        have_normal = (have_normal || normal);
-        have_low = (have_low || low);
+    if ( ! n.pruneWalk() && n.childs().size() ) {
+        auto oprio = prio;
+        for ( const auto& c : n.childs() )
+            prio = std::max(prio, _recursiveValidateAST(c, closest_location, oprio, level + 1, errors));
```

Comment:
I gave this a quick try, but it's not trivial to change the logic so that we don't pass the priority down. I agree this could be cleaned up, but I'll leave it as it is for now; it's just an internal implementation detail.